test/lit/passes/gufa-refs.wast

```diff
@@ -358,6 +360,50 @@
       (local.get $z)
     )
   )
+
+  ;; CHECK:      (func $nondeterminism (type $none_=>_funcref) (result funcref)
+  ;; CHECK-NEXT:  (block $label$1 (result funcref)
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (br_on_func $label$1
+  ;; CHECK-NEXT:     (i31.new
+  ;; CHECK-NEXT:      (i32.const 1337)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (ref.null nofunc)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $nondeterminism (result funcref)
+    ;; This block is sent an i31 and a null. The null is compatible with the
+    ;; type and the i31 is not. The order in which we process this matters:
+    ;;
+    ;;  * If the i31 arrives first, we'll filter it out as incompatible with the
+    ;;    type of the block. Then the null arrives and the final result is that
+    ;;    null, which we can then optimize the block to return.
+    ;;  * Or, if the null arrives first, then when the i31 arrives the
+    ;;    combination of nullfunc + i31 is Many (since the types are
+    ;;    incompatible). We then filter that to the block's type, ending up with
+    ;;    a cone of funcref. We cannot optimize in that case, unlike before.
+    ;;
+    ;; Ideally we'd optimize here, but atm we do not since the order in
+    ;; practice is a less ideal one. At minimum we should be deterministic in
+    ;; how we handle this, which this test enforces at least.
+    ;;
+    ;; TODO: Find a way to actually optimize such cases, perhaps by filtering
+    ;;       when sending as well and not just when receiving (the br_on_func
+    ;;       here should not send anything, as what it sends should be first
+    ;;       intersected with funcref).
```

Comment:
This TODO seems like the better solution. Would it be complicated to implement?

Reply:
It would add work, but even so it does not address the underlying issue I think (see below).

Reply:
Wouldn't this address the underlying issue by making it so that GUFA never has to take any intersections after taking unions, so it would no longer depend on distributivity to make the results deterministic for any iteration order?

Reply:
I still don't understand why this wouldn't solve the underlying issue.

Reply:
Well, again, it can solve the current issue. But there is a fundamental problem when a system doesn't have nice things like distributivity and associativity (sorry if those aren't the right terms in English, but you know what I mean). It's very easy for us to implicitly depend on them later and cause nondeterminism.

I think the only true solutions for the underlying issue are:

1. Get those nice properties. But that's impossibly slow.
2. Order the flow so even without those properties we are fully deterministic.

Reply:
We should have commutativity and associativity (and if we don't, that can be treated as a bug by itself), and I believe that should be enough to make the output independent of the iteration order if we implement this TODO since the rest of the system would never depend on distributivity, which is the only property we don't have.

Reply:
Acknowledged that this belief would be difficult to verify, though 👍

Reply:
Yeah, I sort of think the same, but it's just hard to verify. That this even depends on distributivity is itself kind of subtle - that's only apparent when comparing multiple possible execution paths.

Doing this TODO might make sense. But I'd need to see if it doesn't slow us down. We'd be doing 2 intersections + one merge in each flow iteration instead of 1 of each, and adding the first intersection when I did that definitely added overhead, unfortunately (but it was necessary).